utils/src/main/java/com/cloud/utils/UriUtils.java

```diff
@@ -289,6 +297,93 @@ public static long getRemoteSize(String url) {
+            for (int i = 0; i < tagNames.length; i++) {
+                NodeList targetNodes = rootElement.getElementsByTagName(tagNames[i]);
+                if (targetNodes.getLength() <= 0) {
+                    s_logger.error("no " + tagNames[i] + " tag in XML response...returning null");
+                } else {
+                    List<String> valueList = new ArrayList<String>();
+                    for (int j = 0; j < targetNodes.getLength(); j++) {
+                        Node node = targetNodes.item(j);
+                        valueList.add(node.getTextContent());
+                    }
+                    returnValues.put(tagNames[i], valueList);
+                }
+            }
+        } catch (Exception ex) {
+            s_logger.error(ex);
+        }
+        return returnValues;
+    }
+
+    /**
+     * Check if there is at least one existent URL defined on metalink
+     * @param url metalink url
+     * @return true if at least one existent URL defined on metalink, false if not
+     */
+    protected static boolean checkUrlExistenceMetalink(String url) {
+        HttpClient httpClient = new HttpClient(new MultiThreadedHttpConnectionManager());
+        GetMethod getMethod = new GetMethod(url);
+        try {
+            if (httpClient.executeMethod(getMethod) == HttpStatus.SC_OK) {
+                InputStream is = getMethod.getResponseBodyAsStream();
+                Map<String, List<String>> metalinkUrls = getMultipleValuesFromXML(is, new String[] {"url"});
+                if (metalinkUrls.containsKey("url")) {
+                    List<String> urls = metalinkUrls.get("url");
+                    boolean validUrl = false;
+                    for (String u : urls) {
+                        if (url.endsWith("torrent")) {
+                            continue;
+                        }
+                        try {
+                            UriUtils.checkUrlExistence(u);
+                            validUrl = true;
+                            break;
+                        }
+                        catch (IllegalArgumentException e) {
+                            s_logger.warn(e.getMessage());
+                        }
+                    }
+                    return validUrl;
+                }
+            }
+        } catch (IOException e) {
+            s_logger.warn(e.getMessage());
+        }
+        return false;
+    }
+
+    /**
+     * Get list of urls on metalink
+     * @param metalinkUrl
+     * @return
+     */
+    public static List<String> getMetalinkUrls(String metalinkUrl) {
+        HttpClient httpClient = new HttpClient(new MultiThreadedHttpConnectionManager());
+        GetMethod getMethod = new GetMethod(metalinkUrl);
+        List<String> urls = new ArrayList<>();
+        try {
+            if (httpClient.executeMethod(getMethod) == HttpStatus.SC_OK) {
+                InputStream is = getMethod.getResponseBodyAsStream();
```

Comment:
same remark for the try with resource statement here. Currently the inputstream close is missing.